src/compiler/csharp_generator.cc

```diff
@@ -735,14 +738,15 @@ void GenerateBindServiceWithBinderMethod(Printer* out,
   for (int i = 0; i < service->method_count(); i++) {
     const MethodDescriptor* method = service->method(i);
     out->Print(
-        "serviceBinder.AddMethod($methodfield$, serviceImpl == null ? null : "
+        "serviceBinder.AddMethod($methodfield$, serviceImpl == null ? "
+        "null$nullforgiving$ : "
```

Comment:
It may be worth adding a comment for why this is necessary. (I'm about to look to work it out...)

Reply:
Having looked at the code, I still don't understand why this is necessary. Grpc.Core.Api doesn't have nullable annotations yet anyway, does it? So I'd expect the generated code to be aware of that, and not warn - and when Grpc.Core.Api *does* add nullable annotations, I'd expect ServiceBinderMase.AddMethod to make the second parameter nullable. But I may be missing something...

Reply:
If I remove the null forgiving operator the warning is:
`CS8604 Possible null reference argument for parameter 'handler' in 'void ServiceBinderBase.AddMethod<Point, Feature>(Method<Point, Feature> method, UnaryServerMethod<Point, Feature> handler)'.`
so it does appear to be necessary.

Reply:
We should work out why that is, then. Because it clearly doesn't actually need to be non-null - or we shouldn't be passing in a null reference.

Reply:
The generated code says it is part of an experimental API used in customising service bind logic - not sure if/when it is used:
```C#
 /// <summary>Register service method with a service binder with or without implementation. Useful when customizing the service binding logic.
 /// Note: this method is part of an experimental API that can change or be removed without any prior notice.</summary>
 /// <param name="serviceBinder">Service methods will be bound by calling <c>AddMethod</c> on this object.</param>
 /// <param name="serviceImpl">An object implementing the server-side handling logic.</param>
 [global::System.CodeDom.Compiler.GeneratedCode("grpc_csharp_plugin", null)]
 public static void BindService(grpc::ServiceBinderBase serviceBinder, RouteGuideBase serviceImpl)
 {
   // NOTE - the "null" here needs the null forgiving operation
   serviceBinder.AddMethod(__Method_GetFeature, serviceImpl == null ? null : new grpc::UnaryServerMethod<global::Routeguide.Point, global::Routeguide.Feature>(serviceImpl.GetFeature));
   // ... etc
 }
```

This isn't the `BindService` method that is normally used in Grpc.Core examples - not sure about Grpc.AspNetCore.

Reply:
FYI - it is called from Grpc.AspNetCore via reflection:
https://github.com/grpc/grpc-dotnet/blob/2b1da779bf45a0c55006bc3d36d1e54485c8e834/src/Grpc.AspNetCore.Server/Model/Internal/BinderServiceModelProvider.cs#L54C23-L54C23
```C#
 bindMethodInfo.Invoke(null, new object?[] { binder, null });
```

So I think keeping my fix of adding null forgiving operator is correct.

Reply:
I'm still confused... fundamentally, it sounds like the method really *should* have a nullable parameter, given that we're intentionally calling it with a null argument. So can we fix that? Or have I completely missed something?

(I'm somewhat unnerved by generating calls to an experimental API within a non-experimental generator, admittedly...)

Reply:
The `AddMethod` definitions are in `ServiceBinderBase.cs` in `Grpc.Core.Api`.

Maybe we should add nullable annotations to `Grpc.Core.Api` - but then the generated code is dependent on a particular version of the Grpc.Core.Api package if you don't want to see warnings.

Reply:
I think the safest thing (for now) is to add the annotations to the generated code, then update the code in `Grpc.Core.Api` and  `Grpc.AspNetCore.Server` later.

- If we update `Grpc.Core.Api` and add annotations to `ServiceBinderBase.cs` to say the `handler` parameter may be null then we also need to add annotations to `ProviderServiceBinder.cs` in `Grpc.AspNetCore.Server` as the base class definition will have changed and warnings are treated as errors in that project

- If we don't add the null forgiving operator annotations to the generated code then projects using new code generation but using older `Grpc.Core.Api` without the changes will have warnings.

Reply:
The `!` is required because the `ServiceBinderBase.AddMethod` overloads aren't annotated to allow a nullable handler: https://github.com/grpc/grpc-dotnet/blob/3f908444494b01eed9ffa9f141bbe9f170aee7f6/src/Grpc.Core.Api/ServiceBinderBase.cs#L43

This should be fixed in Grpc.Core.Api. However, even if fixed there, nullable generated code would require a fixed version of Grpc.Core.Api. If an older Grpc.Core.Api version is used, then there would be a warning that null can't be passed to the method.

What I think we should do:
1. Fix `ServiceBinderBase.AddMethod` to make handler nullable.
1. Leave the `!` here but add a comment explaining why it is needed and add a TODO that can eventually be removed. Once the updated version of Grpc.Core.Api has been public for some time, maybe in a years time, then the workaround can be removed.

This would reduce the chance of people seeing warnings from generated code and give a path to a clean state.

Reply:
https://github.com/grpc/grpc-dotnet/pull/2270

Reply:
> The `!` is required because the `ServiceBinderBase.AddMethod` overloads aren't annotated to allow a nullable handler: https://github.com/grpc/grpc-dotnet/blob/3f908444494b01eed9ffa9f141bbe9f170aee7f6/src/Grpc.Core.Api/ServiceBinderBase.cs#L43
> 
> This should be fixed in Grpc.Core.Api. However, even if fixed there, nullable generated code would require a fixed version of Grpc.Core.Api. If an older Grpc.Core.Api version is used, then there would be a warning that null can't be passed to the method.
> 
> What I think we should do:
> 
> 1. Fix `ServiceBinderBase.AddMethod` to make handler nullable in grpc/grpc-dotnet.
> 2. Leave the `!` here, but add a comment explaining why it is needed and add a TODO that it should be removed in the future. Once the updated version of Grpc.Core.Api has been public for some time, maybe in a years time, then the workaround can be removed.
> 
> This would reduce the chance of people seeing warnings from generated code and give a path to a clean state.

I agree with @JamesNK here. Also this is basically what @tonydnewell  has already proposed so we seem to be in agreement.

@tonydnewell  can you please add the comment?